### PR TITLE
Optimize dependency list

### PR DESCRIPTION
**Pull request by latekvo**
## Description

- adds reanimated as a peer dependency of gesture handler.
- narrows down the `react-native` peer dependency version to the last supported one
- removes unnecessary `expo` package
- removes unnecessary `prop-types` package
- removes unused `react-dom` library
- removes multiple unused `@babel` packages
- removes multiple unused testing packages`

## Test plan

- open example app, see how everything works and no errors are thrown
  - [x] test common example (ios, web, android)
  - [x] test MacOS example
  - [x] test Fabric example (ios, android)
- build package, use `rngh` from this `PR` in a fresh react native app
  - [x] test on web
  - [x] test on android
  - [x] test on iOS

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -66,51 +66,44 @@
...
-    "react-native-web": "^0.11.7",
     "react-test-renderer": "18.2.0",
     "release-it": "^13.6.5",
     "typescript": "5.0.4"
   },
   "peerDependencies": {
     "react": "*",
-    "react-native": "*"
+    "react-native": ">= 0.73.0",
+    "react-native-reanimated": ">= 2.0.0"
```
Comment (j-piasecki) [inline package.json:106]: I wouldn't change that - we don't want to have any dependencies on Reanimated (if someone tries to use RNGH with Reanimated 1 it's fine, they just won't integrate). And we don't want to set the lower bound on RN version either - if a new version of RNGH works with the old version of RN it's fine, we just don't guarantee it.

**Review by j-piasecki — COMMENTED** [review]
Do the js tests still pass?

File: package.json (hunk 1 above)
Comment (latekvo) [reply package.json:106]: These dependencies are completely optional, `peerDependencies` don't force the installation of anything and [only alert that the current package versions are not compatible, or that some optional packages are missing](https://github.com/npm/npm/blob/2e3776bf5676bc24fec6239a3420f377fe98acde/doc/files/package.json.md#peerdependencies). 

To be sure this is the case, I `npm pack`ed this PR and installed it in a fresh application:

<details>
<summary>
Collapsed RNGH Installation
</summary>

```bash
ignacylatka:~/dev/ExpoTest52 (main*) $ yarn add ../react-native-gesture-handler/react-native-gesture-handler-2.21.1.tgz
yarn add v1.22.22
[1/4] 🔍  Resolving packages...
[2/4] 🚚  Fetching packages...
[3/4] 🔗  Linking dependencies...
warning "react-native > @react-native/codegen@0.76.1" has unmet peer dependency "@babel/preset-env@^7.1.6".
warning "expo > babel-preset-expo > @react-native/babel-preset > @react-native/babel-plugin-codegen > @react-native/codegen > jscodeshift@0.14.0" has unmet peer dependency "@babel/preset-env@^7.1.6".
warning " > react-native-gesture-handler@2.21.1" has unmet peer dependency "react-native-reanimated@>= 2.0.0".
warning "jest-expo > react-server-dom-webpack@19.0.0-rc-6230622a1a-20240610" has incorrect peer dependency "react@19.0.0-rc-6230622a1a-20240610".
warning "jest-expo > react-server-dom-webpack@19.0.0-rc-6230622a1a-20240610" has incorrect peer dependency "react-dom@19.0.0-rc-6230622a1a-20240610".
warning "jest-expo > react-server-dom-webpack@19.0.0-rc-6230622a1a-20240610" has unmet peer dependency "webpack@^5.59.0".
[4/4] 🔨  Building fresh packages...

success Saved 1 new dependency.
info Direct dependencies
└─ react-native-gesture-handler@2.21.1
info All dependencies
└─ react-native-gesture-handler@2.21.1
✨  Done in 5.04s.
```

</details>

Then if we check if `reanimated` is installed, `yarn why` says it's not:


<details>
<summary>
Collapsed `yarn why` output
</summary>

```bash
ignacylatka:~/dev/ExpoTest52 (main*) $ yarn why react-native-reanimated
yarn why v1.22.22
[1/4] 🤔  Why do we have the module "react-native-reanimated"...?
[2/4] 🚚  Initialising dependency graph...
[3/4] 🔍  Finding dependency...
error We couldn't find a match!
✨  Done in 0.15s.
```

</details>

Every following `yarn install` looks like this:

<details>
<summary>
Collapsed `yarn install` output
</summary>

```bash
ignacylatka:~/dev/ExpoTest52 (main*) $ yarn add @react-navigation/elements                                                                             255
yarn add v1.22.22
[1/4] 🔍  Resolving packages...
[2/4] 🚚  Fetching packages...
[3/4] 🔗  Linking dependencies...
warning "react-native > @react-native/codegen@0.76.1" has unmet peer dependency "@babel/preset-env@^7.1.6".
warning "expo > babel-preset-expo > @react-native/babel-preset > @react-native/babel-plugin-codegen > @react-native/codegen > jscodeshift@0.14.0" has unmet peer dependency "@babel/preset-env@^7.1.6".
warning " > react-native-gesture-handler@2.21.1" has unmet peer dependency "react-native-reanimated@>= 2.0.0".
warning "jest-expo > react-server-dom-webpack@19.0.0-rc-6230622a1a-20240610" has incorrect peer dependency "react@19.0.0-rc-6230622a1a-20240610".
warning "jest-expo > react-server-dom-webpack@19.0.0-rc-6230622a1a-20240610" has incorrect peer dependency "react-dom@19.0.0-rc-6230622a1a-20240610".
warning "jest-expo > react-server-dom-webpack@19.0.0-rc-6230622a1a-20240610" has unmet peer dependency "webpack@^5.59.0".
warning " > @react-navigation/elements@2.1.0" has incorrect peer dependency "@react-navigation/native@^7.0.3".
[4/4] 🔨  Building fresh packages...
success Saved lockfile.
success Saved 1 new dependency.
info Direct dependencies
└─ @react-navigation/elements@2.1.0
info All dependencies
└─ @react-navigation/elements@2.1.0
✨  Done in 2.47s.
```

</details>

The following line is the only difference `peerDependencies` make: 
```bash
warning " > react-native-gesture-handler@2.21.1" has unmet peer dependency "react-native-reanimated@>= 2.0.0".
```

I think having both Reanimated, and the oldest known compatible version of React Native correctly utilises what `peerDependencies` was made for. 
It's a more verbose way of showing which package versions are expected, without forcing their installation. 
The only other places where the compatible package versions are listed are [here](https://docs.swmansion.com/react-native-gesture-handler/docs/fundamentals/installation/#requirements) and [here](https://www.npmjs.com/package/react-native-gesture-handler), and the user has to already know their issue lies in package version incompatibilities to look at these tables.

This change can only ever result in a non-critical warning, and only when the user uses library versions which are [explicitly listed as incompatible](https://docs.swmansion.com/react-native-gesture-handler/docs/fundamentals/installation/#requirements).

Comment (latekvo) [reply package.json:106]: My bad, after looking at the documentation some more, i think you were right.
I think it'll be better to keep `React Native` version at `*`, and either remove `Reanimated`, or move it to `optionalDependencies`, but I'll have to read on them more to make sure `Reanimated` can be included there

Comment (latekvo) [reply package.json]: reverted in [04b1159](https://github.com/software-mansion/react-native-gesture-handler/pull/3216/commits/04b11595bfc731789dbdcbf802db93351e62af9a)

**Comment by j-piasecki** [comment]
I meant `yarn test`

**Comment by latekvo** [comment]
Yes, `yarn test` passes as well

image

**Review by m-bert — APPROVED** [review]

**Pull request merged**